core/deployment/src/main/java/io/quarkus/deployment/steps/ExecutionHandlersBuildStep.java

```diff
@@ -0,0 +1,238 @@
+                ResultHandle features = mv.load(featuresItems.stream()
+                        .map(FeatureBuildItem::getInfo)
+                        .sorted()
+                        .collect(Collectors.joining(", ")));
+                ResultHandle version = mv.load(Version.getVersion());
+                final ResultHandle newCtx = mv.invokeVirtualMethod(WITH_VALUES, mv.getMethodParam(1),
+                        mv.readStaticField(FieldDescriptor.of(ExecutionChain.class, "QUARKUS_VERSION", String.class)),
+                        version,
+                        mv.readStaticField(FieldDescriptor.of(ExecutionChain.class, "QUARKUS_FEATURES", String.class)),
+                        features);
+                final ResultHandle result = mv.invokeVirtualMethod(PROCEED, mv.getMethodParam(0), newCtx);
+                mv.returnValue(result);
+            }
+            reflectiveClasses.accept(new ReflectiveClassBuildItem(true, false, false, VERSION_AND_FEATURES_CLASS_NAME));
+            return new ExecutionHandlerBuildItem(new SupplierExecutionHandler(VERSION_AND_FEATURES_CLASS_NAME));
+        }
+    }
+
+    @BuildStep
+    @Before(QuarkusApplicationBuildItem.class)
+    public void generateMainRecorders(
+            ClassOutputBuildItem outputItem,
+            List<MainBytecodeRecorderBuildItem> mainTasks,
+            List<ObjectSubstitutionBuildItem> substitutions,
+            List<BytecodeRecorderObjectLoaderBuildItem> loaders,
+            Consumer<ReflectiveClassBuildItem> reflectiveClasses,
+            Consumer<ExecutionHandlerBuildItem> executionHandlers) {
+
+        for (MainBytecodeRecorderBuildItem recItem : mainTasks) {
+            final BytecodeRecorderImpl recorder = recItem.getBytecodeRecorder();
+            if (!recorder.isEmpty()) {
+                // Register substitutions in all recorders
+                for (ObjectSubstitutionBuildItem sub : substitutions) {
+                    ObjectSubstitutionBuildItem.Holder holder1 = sub.holder;
+                    recorder.registerSubstitution(holder1.from, holder1.to, holder1.substitution);
+                }
+                for (BytecodeRecorderObjectLoaderBuildItem item : loaders) {
+                    recorder.registerObjectLoader(item.getObjectLoader());
+                }
+                recorder.writeBytecode(outputItem.getClassOutput());
+                executionHandlers.accept(new ExecutionHandlerBuildItem(new SupplierExecutionHandler(recorder.getClassName())));
+                reflectiveClasses.accept(new ReflectiveClassBuildItem(true, false, false, recorder.getClassName()));
+            }
+        }
+    }
+
+    @BuildStep
+    @Before(QuarkusApplicationBuildItem.class)
+    public GeneratedResourceBuildItem serializeExecutionChain(ExecutionChainBuildItem chainBuildItem) throws IOException {
+        ExecutionChain chain = chainBuildItem.getChain();
+
+        // Serialize the chain
```

Comment:
> I am 100% opposed to this approach. I think it would be a huge step backwards, and undermines what we are trying to do with Quarkus.

Do you have evidence to support this opinion?

> The whole point of the current architecture is to do as little as possible at startup time. Adding serialisation to startup is the opposite of 'doing as little as possible at startup'. While this might not add noticeable overhead at the moment and the number of classes is small as time goes on if this is used more often this will increase.

Sure, this can be easily addressed at that time.  Or sooner.

> It also opens the door to other similar changes, all of which are 'fast' when taken in isolation, however over time lots of 'fast' changes add up to slow startup. We can't both have a 'do as little as possible at startup' philosophy and allow changes like this in.

Again, evidence?

> I also do not believe there is any technical reason for this change. It is quite possible to implement command mode applications with the current architecture (e.g. the AWS lambda stuff basically works the same as a command mode application).

The technical reason is simply to materialize the execution chain during static init.  Of course we could replace this mechanism with (for example) bytecode-generated deserialization.  But this takes more time and I wanted to get this out the door, viewing it as a premature optimization for now.

Reply:
> > I am 100% opposed to this approach. I think it would be a huge step backwards, and undermines what we are trying to do with Quarkus.
> 
> Do you have evidence to support this opinion?

Because the whole architecture is based around doing as little as possible on startup to reduce startup time as much as possible. Adding extra work on startup that is not technically necessary is a step backwards.

> 
> > The whole point of the current architecture is to do as little as possible at startup time. Adding serialisation to startup is the opposite of 'doing as little as possible at startup'. While this might not add noticeable overhead at the moment and the number of classes is small as time goes on if this is used more often this will increase.
> 
> Sure, this can be easily addressed at that time. Or sooner.
> 
> > It also opens the door to other similar changes, all of which are 'fast' when taken in isolation, however over time lots of 'fast' changes add up to slow startup. We can't both have a 'do as little as possible at startup' philosophy and allow changes like this in.
> 
> Again, evidence?

It's human nature. If we say yes to this then its harder to say no to similar changes that add extra work to startup.

> 
> > I also do not believe there is any technical reason for this change. It is quite possible to implement command mode applications with the current architecture (e.g. the AWS lambda stuff basically works the same as a command mode application).
> 
> The technical reason is simply to materialize the execution chain during static init. Of course we could replace this mechanism with (for example) bytecode-generated deserialization. But this takes more time and I wanted to get this out the door, viewing it as a premature optimization for now.

Reply:
> > > I am 100% opposed to this approach. I think it would be a huge step backwards, and undermines what we are trying to do with Quarkus.
> > 
> > 
> > Do you have evidence to support this opinion?
> 
> Because the whole architecture is based around doing as little as possible on startup to reduce startup time as much as possible. Adding extra work on startup that is not technically necessary is a step backwards.

Only data will show if it's a step backwards.

> > Again, evidence?
> 
> It's human nature. If we say yes to this then its harder to say no to similar changes that add extra work to startup.

This doesn't really prove anything.  I have a full work load so I'm not really interested in having this kind of debate.  Point to a part of the code that means the change cannot be merged, with a technical justification for that assertion, and we'll discuss.